include/aws/common/priority_queue.h

```diff
@@ -0,0 +1,97 @@
+#ifndef AWS_PRIORITY_QUEUE_H
+#define AWS_PRIORITY_QUEUE_H
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <aws/common/array_list.h>
+
+typedef int(*aws_priority_queue_compare)(void *a, void *b);
```

Comment:
Ergonomics: is `aws_priority_queue_` too long for a prefix?

Reply:
Also, please precisely define the semantics of this operation. The semantics appear to differ from normal C comparator semantics (which is probably a bad idea). Ideally this should be a comparator that would also work in, say, qsort.

The pointers should also be `const void *`.

Reply:
I did document the semantics. https://github.com/awslabs/aws-c-common/pull/7/files#diff-dc4707d66f9b3bc969cb897313db7982R25

It's not the same semantics of a normal C comparator. In this case we need a predicate not the full comparison that returns negative/0/positive.

Reply:
True, but if we define "The comparator should return a positive value if `a` should be returned before `b`; if `b` should come before `a` or the order is unspecified, it should return either a negative value or zero" then the semantics are defined precisely but are also compatible with generic sort routines.

Reply:
> Ergonomics: is aws_priority_queue_ too long for a prefix?

Not with a good text editor 😛 
>The pointers should also be const void *.

Will change it.

>The comparator should return a positive value if a should be returned before b; if b should come before a or the order is unspecified, it should return either a negative value or zero"

Fine, I'll change them. But note that, like in std::priority_queue providing std::less<> results in a max heap. So, I'll reword what you suggested here to fit the same semantics while making it compatible with qsort

Reply:
Rewording is fine, I'm just trying to avoid people shooting themselves in the foot when they accidentally pass a qsort-compatible function to this

Reply:
updated. And you're right. It's better this way.